### PR TITLE
PT-28 Use webpack instead of gulp to compile project on boilerplate

### PR DESCRIPTION
**Pull request by akashlakhera**
## Description

Use webpack instead of gulp to compile boilerplate.

## Screenshots

https://user-images.githubusercontent.com/17554101/192758444-a393d12b-240e-4a66-b78f-7b68edd87524.mov

### CODE REVIEW

**Comment by bbbneo333** [comment]
@akashlakhera - Encountered following issue when ran `docker compose -f docker-compose.dev.yml up --build`, can you please check?

Screenshot 2022-09-29 at 12 04 42 PM

**Comment by bbbneo333** [comment]
Hey @akashlakhera, while testing out the new build system, noticed that when there's is change in backend, it tries to build the frontend as well (please see video, atleast according to the logs it looks like it but might be wrong). 

https://user-images.githubusercontent.com/23015711/193202100-dc8e370f-94a9-4a09-ab69-bbf5e57f52be.mov

File: config/webpack/webpack.prod.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (bbbneo333) [inline config/webpack/webpack.prod.js:1]: See note from above

File: config/webpack/webpack.prod.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+// eslint-disable-next-line no-unused-vars
+const path = require('path');
```
Comment (bbbneo333) [inline config/webpack/webpack.prod.js:3]: Think these can be removed

**Review by bbbneo333 — CHANGES_REQUESTED** [review]

File: config/webpack/webpack.common.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      },
+      {
+        test: /\.ts(x)?$/,
+        loader: 'ts-loader',
+        exclude: /node_modules/,
+      },
+    ],
+  },
+  plugins: [
+    new webpack.DefinePlugin({ CONFIG: JSON.stringify(CONFIG) }),
```
[the remark this answers is not in the document]
Comment (akashlakhera) [reply config/webpack/webpack.common.js:44]: This should not inject entire configuration according to [docs](https://github.com/node-config/node-config/wiki/Webpack-Usage#option-1).

[the remark this answers is not in the document]
Comment (bbbneo333) [reply config/webpack/webpack.common.js:44]: Still, don't think this is still something we should be doing. As mentioned:
- Define a `public` config which is dedicated to frontend. Nothing else should injected (irrespective of env)
- The type we have defined for config (`const CONFIG: { [key: string]: { [key: string]: string } };`) should be explicit. 

Say if we have:
```json
{
  "public": {
    "some_config": "some value"
  }
}
```

We should have type definition as:
```typescript
type Config = {
  'some_config': string
};
```

And we should be able to use it as `Config.some_config` in frontend.

File: src/public/js/ui/app.tsx [hunk 4]
```diff
@@ -6,6 +6,7 @@ import InspectLet from '../vendor/inspectlet';
 
 export const App: React.FC = () => {
   useEffect(() => {
+    console.log(CONFIG.server.port);
```
Comment (bbbneo333) [inline src/public/js/ui/app.tsx:9]: Can show the current env on homepage instead

**Pull request closed without merging**